### PR TITLE
feat(dialect): snowflake dialect support

### PR DESCRIPTION
**Pull request by jesse23**
### Pull Request check-list

- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Have you added new tests to prevent regressions?
- [x] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Did you update the typescript typings accordingly (if applicable)?
- [x] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)?

### Description of change

snowflake dialect support.

- Supported by [Dynatrace](https://www.dynatrace.com) and [Snowflake](http://www.snowflake.com)
- Contributed by [jesse23](https://github.com/jesse23), [josecolella](https://github.com/josecolella), [cmp11290](https://github.com/cmp11290), [cincyelite22](github.com/cincyelite22), [jdanielmyers](https://github.com/jdanielmyers)

### Note
- All our use cases has been covered by unit test and integration test.
- We have limited use cases so yes it might not have 100% coverage - that could be improved by follow up PR if possible 

### How to run integration test
- Snowflake account is required
- Run command below:
```
DIALECT=snowflake SEQ_ACCOUNT=<snowflake-account.us-east-1> SEQ_USER=<snowflake-user-name> SEQ_PW=<snowflake-password> SEQ_ROLE=<snowflake-role> SEQ_DB=<DB_NAME> SEQ_SCHEMA=<DV_SCHEMA> SEQ_WH=<WHAREHOUSE_NAME> npm run test-integration
```

### Link with existing defect:
https://github.com/sequelize/sequelize/issues/11979

### CODE REVIEW

**Comment by jesse23** [comment]
Pipeline is green.... Let me know if any other actions needed...thx @Keimeno

File: lib/dialects/snowflake/query-generator.js
Comment (AllAwesome497) [inline lib/dialects/snowflake/query-generator.js:167]: nit: use Utils.joinSQLFragments.

Something like:
```js
return Utils.joinSQLFragments([
  'part1',
  database 
    ? part2_a
    : part2_b,
  ';'
]);
```

File: lib/dialects/snowflake/query-generator.js
Comment (AllAwesome497) [inline lib/dialects/snowflake/query-generator.js:238]: nit: use `Utils.joinSQLFragments`

File: lib/dialects/snowflake/query-generator.js
Comment (AllAwesome497) [inline lib/dialects/snowflake/query-generator.js:258]: I have no idea what `smth` is - can you rename and/or just explain?

File: lib/dialects/snowflake/query-generator.js
Comment (AllAwesome497) [inline lib/dialects/snowflake/query-generator.js:315]:
```suggestion
```

File: lib/dialects/snowflake/query-generator.js
Comment (AllAwesome497) [inline lib/dialects/snowflake/query-generator.js:321]: nit: change this to use `Utils.joinSQLFragments`.

File: lib/dialects/snowflake/query-generator.js
Comment (AllAwesome497) [inline lib/dialects/snowflake/query-generator.js:385]: `Utils.joinSQLFragments` could also be used here

File: lib/dialects/snowflake/query-generator.js
Comment (AllAwesome497) [inline lib/dialects/snowflake/query-generator.js:549]: I don't think trust this this regex - there's any other `)` after the closing one everything up to that will be replaced.

File: lib/dialects/snowflake/query-generator.js
Comment (AllAwesome497) [inline lib/dialects/snowflake/query-generator.js:633]:
```suggestion
module.exports = SnowflakeQueryGenerator;
```

**Review by AllAwesome497 — CHANGES_REQUESTED** [review]
Haven't looked at everything, but here's some feedback for now.

**Comment by jesse23** [comment]
@AllAwesome497  Thx! Yes it is fork from mySQL Dialect and most of use cases we are using are adopted and tested in snowflake integration test....

I will fix most of the missing name changing and ping u back.

Also regarding to `extend the mysql query generator` I will take a look and let u know. 

Thx again!

File: lib/dialects/snowflake/connection-manager.js
Comment (jesse23) [reply lib/dialects/snowflake/connection-manager.js:91]: That is not supported by snowflake - it only supports string like `us` for now.

File: lib/dialects/snowflake/data-types.js
Comment (jesse23) [reply lib/dialects/snowflake/data-types.js:82]: addressed.

File: lib/dialects/snowflake/query-generator.js
Comment (jesse23) [reply lib/dialects/snowflake/query-generator.js:321]: yes

File: lib/dialects/snowflake/query-generator.js
Comment (jesse23) [reply lib/dialects/snowflake/query-generator.js:385]: `attributeToSQL` better to stay in existing...it is a bit complex

File: lib/dialects/snowflake/connection-manager.js
Comment (jesse23) [reply lib/dialects/snowflake/connection-manager.js:43]: snowflake sql looks like a mix between mysql and postsql to me....we can keep it separate for now and refactor it after we have most of the use case coverage.

**Comment by jesse23** [comment]
@AllAwesome497 All comments addressed. Thx!

File: lib/dialects/snowflake/data-types.js
Comment (jesse23) [reply lib/dialects/snowflake/data-types.js]: done

File: lib/dialects/snowflake/query-generator.js
Comment (jesse23) [reply lib/dialects/snowflake/query-generator.js:637]: addressed

File: lib/dialects/snowflake/query-generator.js
Comment (jesse23) [reply lib/dialects/snowflake/query-generator.js:256]: This is used acorss the code base...I will rename it to `attr` in snowflake branch

File: lib/dialects/snowflake/query-generator.js
Comment (jesse23) [reply lib/dialects/snowflake/query-generator.js:234]: better to stay existing for now...the logic is different with `Utils.joinSQLFragments` with auto adding space

File: lib/dialects/snowflake/query-interface.js
Comment (AllAwesome497) [inline lib/dialects/snowflake/query-interface.js:58]: small nit: For things like this (single flags) I think you can just do this in a single line, which imo looks neater.
```suggestion
  /** @override */
  async removeConstraint(tableName, constraintName, options) {
```

File: lib/dialects/snowflake/query.js
Comment (AllAwesome497) [inline lib/dialects/snowflake/query.js:60]: Should this be using the same logging system as everything else?

**Review by AllAwesome497 — COMMENTED** [review]
Still haven't looked through everything, but here's some more feedback.

Ideally this would be broken up into multiple prs - e.g. `connection manager` -> `queryGenerator` -> `queryInterface`.

**Comment by jesse23** [comment]
> Still haven't looked through everything, but here's some more feedback.
> 
> Ideally this would be broken up into multiple prs - e.g. `connection manager` -> `queryGenerator` -> `queryInterface`.

Thx will try to catch it up today or tomorow.

Regarding to the PR split, it might be good to promote something working (with the real server) as 1st commit...sorry for the heavy load of review.

**Comment by AllAwesome497** [comment]
> Regarding to the PR split, it might be good to promote something working (with the real server) as 1st commit

Agreed.  There should probably be a process for adding dialects added to the contributing docs.

Ideally this would be broken up into multiple prs but since there isn't any established guidelines for this it's probably fine as is.

File: lib/dialects/mysql/data-types.js
Comment (jesse23) [reply lib/dialects/mysql/data-types.js:57]: Sorry...yes by mistake

File: lib/dialects/snowflake/query-interface.js
Comment (jesse23) [reply lib/dialects/snowflake/query-interface.js:58]: This practice is from other dialect - but yes I will follow ur suggestion in snowflake dialect.

File: lib/dialects/snowflake/query.js
Comment (jesse23) [reply lib/dialects/snowflake/query.js:18]: Agree - will remove it and let us see ( this is also from practice in other dialect)

File: lib/dialects/snowflake/query.js
Comment (jesse23) [reply lib/dialects/snowflake/query.js:60]: Yes - sorry for the debug left over...will remove it and rely on framework promise processing.

File: lib/dialects/snowflake/query-generator.js
Comment (jesse23) [reply lib/dialects/snowflake/query-generator.js:309]: Yes this is the credit after using joinSQLFragments

File: lib/dialects/snowflake/query-generator.js
Comment (jesse23) [reply lib/dialects/snowflake/query-generator.js:234]: This `changeColumnQuery` is close to the postgres when other part of are more close to mysql....

I will refactor it to use `joinSQLFragments`

Comment (jesse23) [reply lib/dialects/snowflake/query-generator.js:234]: Done - use `joinSQLFragens` and also remove the ENUM code ( since snowflake doesn't support it )

**Comment by jesse23** [comment]
@AllAwesome497  All addressed - thanks for the review! 

Let me know anything needs to be update and fixed.

**Comment by jesse23** [comment]
@Keimeno @AllAwesome497 Hi could we have another review and proceed this PR? Thanks...

**Comment by AllAwesome497** [comment]
> @Keimeno @AllAwesome497 Hi could we have another review and proceed this PR? Thanks...

Haven't taken the time to look at sequelize prs since there isn't any active core maintainers- so even if its approved, I don't see it getting merged any time soon.  I'll let you know if the situation changes.

**Comment by sdepold** [comment]
@jesse23 Could you please ping me on slack?

**Comment by sdepold** [comment]
We are going to have our monthly sequelize sync tomorrow and will discuss the matter of how to deal with dialect support requests there: https://www.meetup.com/eBay-Europe-Technology/events/282125063/

Feel free to drop by and participate in the discussion.

**Comment by jesse23** [comment]
> We are going to have our monthly sequelize sync tomorrow and will discuss the matter of how to deal with dialect support requests there: https://www.meetup.com/eBay-Europe-Technology/events/282125063/
> 
> Feel free to drop by and participate in the discussion.

Thanks - yes will attend. FYI @cincyelite22

File: lib/dialects/abstract/query-generator/helpers/quote.js
Comment (sdepold) [inline lib/dialects/abstract/query-generator/helpers/quote.js:30]: Would be really cool if this could be moved over to the snowflake specific dialect files

File: test/unit/sql/create-table.test.js
Comment (sdepold) [inline test/unit/sql/create-table.test.js:11]: Can we do something like

```
if(dialect=snowflake){
return
}
```

?

File: test/unit/sql/data-types.test.js
Comment (sdepold) [inline test/unit/sql/data-types.test.js:77]: the default seems to be text already

File: test/unit/sql/index.test.js
Comment (sdepold) [inline test/unit/sql/index.test.js:12]: please return early to prevent the huge diff

**Review by sdepold — CHANGES_REQUESTED** [review]

**Comment by jesse23** [comment]
@cincyelite22 Please follow up. Thx.

File: lib/dialects/abstract/query-generator/helpers/quote.js
Comment (cincyelite22) [reply lib/dialects/abstract/query-generator/helpers/quote.js:30]: @sdepold This is currently following the same practice as what postgres is doing for reserved words. Do you have a recommendation on where this can neatly be injected without refactoring all of the calls to this function?

Comment (jesse23) [reply lib/dialects/abstract/query-generator/helpers/quote.js:30]: @sdepold I will follow up in the dialect separation work with postgres together.

File: .github/workflows/ci.yml
Comment (Keimeno) [inline .github/workflows/ci.yml:209]: Please change the highest tested node version from 12 to 16, we recently changed this

**Comment by sdepold** [comment]
Docs seems to be broken now. But we are getting closer :)

**Comment by jmikolay22** [comment]
Should be ready now.

**Review by sdepold — APPROVED** [review]

**Comment by sdepold** [comment]
I think you'll have to update with master one more time

**Comment by josecolella** [comment]
🥳

**Pull request merged**
